activerecord/lib/active_record/migration.rb

```diff
@@ -1157,5 +1186,29 @@ def ddl_transaction(migration)
     def use_transaction?(migration)
       !migration.disable_ddl_transaction && Base.connection.supports_ddl_transactions?
     end
+
+    def use_advisory_lock?
+      Base.connection.supports_advisory_locks?
+    end
+
+    def with_advisory_lock
+      key = generate_migrator_advisory_lock_key
+      got_lock = Base.connection.get_advisory_lock(key)
+      raise ConcurrentMigrationError unless got_lock
+      load_migrated # reload schema_migrations to be sure it wasn't changed by another process before we got the lock
+      yield
+    ensure
+      Base.connection.release_advisory_lock(key) if got_lock
+    end
+
+    # Postgres limits us to 64 bit signed integers
+    # MySQL advisory locks are server-wide so we have to scope to the database
+    # MIGRATOR_SALT is a a random signed integer (max 31 bits) that identifies
+    # the Rails migration process as the owner of the lock.
```

Comment:
Again, these sort of implementation detail comments tend to get out of sync. Let's just leave this in the commit message.